File: client/ios/App/Navigation/MobileLinkParser.swift
import Foundation

// Marketing America Corp. Oleksandr Tishchenko
public enum MobileLinkParser {
    private static let customPrefix = "smartresponsor://mobile/"
    private static let universalPrefix = "https://app.smartresponsor.com/mobile/"

    public static func parse(_ raw: String?) -> MobileLink? {
        let value = (raw ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
        guard !value.isEmpty else {
            return nil
        }

        let normalized: String
        if value.hasPrefix(customPrefix) {
            normalized = String(value.dropFirst(customPrefix.count))
        } else if value.hasPrefix(universalPrefix) {
            normalized = String(value.dropFirst(universalPrefix.count))
        } else if value.hasPrefix("/") {
            normalized = String(value.drop { $0 == "/" })
        } else {
            normalized = value
        }

        let parts = normalized.split(separator: "?", maxSplits: 1, omittingEmptySubsequences: false)
        let route = normalizeRoute(String(parts.first ?? ""))
        guard !route.isEmpty else {
            return nil
        }

        let query = parts.count > 1 ? parseQuery(String(parts[1])) : [:]
        return MobileLink(raw: value, route: route, query: query)
    }

    private static func normalizeRoute(_ value: String) -> String {
        value
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .split(separator: "/", omittingEmptySubsequences: true)
            .joined(separator: "/")
    }

    private static func parseQuery(_ value: String) -> [String: String] {
        var result: [String: String] = [:]
        for part in value.split(separator: "&", omittingEmptySubsequences: true) {
            let pair = part.split(separator: "=", maxSplits: 1, omittingEmptySubsequences: false)
            let key = String(pair.first ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
            guard !key.isEmpty else {
                continue
            }
            result[key] = pair.count > 1 ? String(pair[1]).trimmingCharacters(in: .whitespacesAndNewlines) : ""
        }
        return result
    }
}

File: client/android/app/src/main/java/app/mobiling/client/navigation/MobileLinkParser.kt
package app.mobiling.client.navigation

/**
 * Marketing America Corp. Oleksandr Tishchenko
 */
object MobileLinkParser {
    private const val CUSTOM_PREFIX = "smartresponsor://mobile/"
    private const val UNIVERSAL_PREFIX = "https://app.smartresponsor.com/mobile/"

    fun parse(raw: String?): MobileLink? {
        val value = raw?.trim().orEmpty()
        if (value.isEmpty()) {
            return null
        }

        val normalized = when {
            value.startsWith(CUSTOM_PREFIX) -> value.removePrefix(CUSTOM_PREFIX)
            value.startsWith(UNIVERSAL_PREFIX) -> value.removePrefix(UNIVERSAL_PREFIX)
            value.startsWith("/") -> value.trimStart('/')
            else -> value
        }

        val routeAndQuery = normalized.split("?", limit = 2)
        val route = normalizeRoute(routeAndQuery.firstOrNull().orEmpty())
        if (route.isEmpty()) {
            return null
        }

        return MobileLink(
            raw = value,
            route = route,
            query = routeAndQuery.getOrNull(1)?.let(::parseQuery).orEmpty(),
        )
    }

    private fun normalizeRoute(value: String): String = value
        .trim()
        .trim('/')
        .replace(Regex("/{2,}"), "/")

    private fun parseQuery(value: String): Map<String, String> = value
        .split('&')
        .filter { it.isNotBlank() }
        .mapNotNull { part ->
            val pair = part.split("=", limit = 2)
            val key = pair[0].trim()
            if (key.isBlank()) {
                null
            } else {
                key to pair.getOrNull(1).orEmpty().trim()
            }
        }
        .toMap()
}
